python/paddle/fluid/contrib/slim/quantization/dyquant/dygraph_quantization.py

```diff
@@ -0,0 +1,108 @@
+
+
+class DygraphQuantizationPass(object):
+    def __init__(self,
+                 weight_bits=8,
+                 activation_bits=8,
+                 moving_rate=0.9,
+                 quantizable_layer_type=['Conv2D', 'Linear']):
+        super(DygraphQuantizationPass, self).__init__()
+        self._weight_bits = weight_bits
+        self._activation_bits = activation_bits
+        self._moving_rate = moving_rate
+        self._quantizable_layer_type = quantizable_layer_type
+        self._quant_layers_map = {'Conv2D': Conv2D, 'Linear': Linear}
+        self._translator = ProgramTranslator()
+
+    def prepare_quantize(self):
+        self._translator.enable_declarative = False
+
+    def save_infer_quant_model(self,
+                               dirname,
+                               model,
+                               input_shape,
+                               input_dtype='float32',
+                               feed=None,
+                               fetch=None,
+                               append_batch_size=True):
+        with guard():
+            self._translator.enable_declarative = True
+            model.eval()
+            raw_data = np.random.random(input_shape)
+            input_data = raw_data[np.newaxis, :].astype(
+                input_dtype) if append_batch_size else raw_data.astype(
+                    input_dtype)
+            input_var = to_variable(input_data)
+            out = model(input_var)
+
+        self._translator.save_inference_model(dirname, feed, fetch)
+
+    def quantize_qat(self, model):
+        quant_layers = tuple(self._quant_layers_map[layer]
+                             if layer in self._quant_layers_map else layer
+                             for layer in self._quantizable_layer_type)
+        for layer in quant_layers:
+            assert not isinstance(
+                layer, str), "{} is unspported to be quantized.".format(layer)
+
+        for name, layer in model.named_sublayers():
+            if not isinstance(layer, quant_layers):
+                continue
+
+            scopes = name.split('.')
+            target = scopes[-1]
+            obj = model
+            parent = model
+            for i in range(len(scopes) - 1):
+                obj = getattr(parent, scopes[i])
+                parent = obj
+
+            quant_layer = self._get_quantized_counterpart(layer)
+            setattr(obj, target, quant_layer)
+
+    def _get_quantized_counterpart(self, layer):
+        quant_layers = tuple(self._quant_layers_map.values())
+        quantized_counterpart = tuple('Quantized' + k
+                                      for k in self._quant_layers_map.keys())
+
+        predicate = lambda value: isinstance(layer, value)
+        index_generator = (i for i, v in enumerate(quant_layers)
+                           if predicate(v))
+
+        try:
+            index = next(index_generator)
+        except StopIteration:
```

Comment:
If a model has no quant_layer, can it just be a warning? Should we throw fatal error and exit it?